keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+
+*Further improvements to the testing API*
+
+- Make the function calls in network policy builder *even* more DSL-like, for example, 
+```
+Pod(...).InNamespace(...).CanAccess(...)
+```
+- Infer `Egress,Ingress` rules rather than force them to be specified, based on builder inputs.  They're redundant to begin with (i.e. calico doesn't even require them)
+- Add `From` and `To` semantics to the struct API calls in reachability.
+ 
+###  Note on Acceptance and Backwards compatibility
+
+Thus far there are two obvious ways to ensure backwards compatibility. 
+- Each old test will be converted to a truth table, first, as part of this work, so that the parity between old and new tests is clear and obvious 
+- Alternatively, these can be a next generation of policy tests which live in parallel to existing network policy tests for a release cycle, while they are vetted
+
+## Next steps
+
+As of now, network policy tests are not run regularly against any CNI.  Although we should not endorse one CNI over another, we should regularly validate
+that the NetworkPolicy tests *can* pass on *some* provider.  As part of this proposal, we propose committing an annotation to the existing `network_policy.go` code which states, in clear and simple terms, what environment the `network_policy.go` test suite was run in, the last time which it was committed and passed.  It's also acceptable to commit this as a Markdown file in the documentation.
+ 
+There may be other, better ways of doing this.  Running an upstream validation job of these tests as a weekly PROW job, for example, would be a good way to make sure that these tests don't regress in the future.  This comes at the cost of coupling a job to an external CNI provider, so its not being explicitly suggested.
+
+## Other Improvement Ideas
+
+These may be included in this proposal, but as of now, aren't detailed yet.
+
+### Node local traffic: Should it be revisited ? 
+
+There is an interesting caveat in the types.go API definition for NetworkPolicy.Ingress, wherein we mention that node local traffic is ALLOWED by default.  This is to enable health checks.  Although for now we can add this test, it's worth leveraging some of the thought process here to rethink wether this security hole is wanted long term in the network policy API.
+
+### Validation and 'type' : Should it be revisited ?
+
+There are some corner cases for validation (like type=Ingress, but an existing `egress:` stanza is present) which are not currently validated against.  Should we start rejecting such policies, since after all, they don't make any sense (i.e. sending an egress payload when the type is Ingress has no value, unless you're planning on toggling ingress/egress on and off over time, which seems like would be easier done by simply creating/deleting new policies.  And even if toggling were desired, you could do this with explicit API constructs "EgressEnabled:true", "IngressEnabled:true".
```

Comment:
we cannot start rejecting policies that are not currently rejected

Reply:
fair enough, we can  move this to the Netpol 2 KEP